### PR TITLE
BIGNUM

### PR DESCRIPTION
**Pull request by julianrocha**
Purpose of this PR is mainly to see what I've been doing, not ready to merge yet.

Currently not working on large keys for some reason so I need to debug. It works on small keys right now so it's ready to be tested on the VM for these purposes. I will hold off on merging until bug is fixed and so that I don't clash with the unint96 stuff. Most useful overview file to check out is `bignum.h`

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+	"files.associations": {
+		"bignum.h": "c"
+	}
+}
```
Comment (ryanrussell10) [inline .vscode/settings.json:5]: Probably don't need this in the repo.

File: mock1.c [hunk 2]
```diff
@@ -1,120 +0,0 @@
-#include <stdio.h>
```
Comment (ryanrussell10) [inline mock1.c:1]: This deletion won't affect the uint96_t stuff, since that is in mock2.c

File: bignum.h [hunk 3]
```diff
@@ -25,8 +37,32 @@ typedef struct bignum
...
+void bignum_sub(bignum *a, bignum *b, bignum *c);				/* c = a - b */
+void bignum_mul(bignum *a, bignum *b, bignum *c);				/* c = a * b */
+void bignum_div(bignum *a, bignum *b, bignum *c);				/* c = a / b */
+void bignum_mod(bignum *a, bignum *b, bignum *c);				/* c = a % b */
+void bignum_divmod(bignum *a, bignum *b, bignum *c, bignum *d); /* c = a/b, d = a%b */
+
+/* Bitwise operators: */
+void bignum_rshift(bignum *a, bignum *b, unsigned int nbits); /* b = a >> nbits */
+void bignum_or(bignum *a, bignum *b, bignum *c);			  /* c = a | b */
+
```
Comment (ryanrussell10) [inline bignum.h:58]: Any plans to add support for AND or XOR, or have you determined that they won't be required?

File: bignum.h [hunk 4]
```diff
@@ -25,8 +37,32 @@ typedef struct bignum
...
+/* Arithmetic operations: */
+void bignum_add(bignum *a, bignum *b, bignum *c);				/* c = a + b */
+void bignum_sub(bignum *a, bignum *b, bignum *c);				/* c = a - b */
+void bignum_mul(bignum *a, bignum *b, bignum *c);				/* c = a * b */
+void bignum_div(bignum *a, bignum *b, bignum *c);				/* c = a / b */
+void bignum_mod(bignum *a, bignum *b, bignum *c);				/* c = a % b */
+void bignum_divmod(bignum *a, bignum *b, bignum *c, bignum *d); /* c = a/b, d = a%b */
+
+/* Bitwise operators: */
```
Comment (ryanrussell10) [inline bignum.h:55]: Missing definitions for lshift and the one bit shifts?

File: bignum.c [hunk 5]
```diff
@@ -61,4 +73,248 @@ void print_bignum(bignum *n)
...
+}
+
+static void _rshift_one_bit(bignum *a)
+{
+	for (int i = 0; i < (BN_ARRAY_LEN - 1); ++i)
+	{
+		a->arr[i] = (a->arr[i] >> 1) | (a->arr[i + 1] << ((8 * WORD_BYTES) - 1));
+	}
+	a->arr[BN_ARRAY_LEN - 1] >>= 1;
 }
```
Comment (ryanrussell10) [inline bignum.c:320]: Looks good, thanks for setting this up :)

**Review by ryanrussell10 — APPROVED** [review]
Looks good, thanks for setting this up :)

File: bignum.h (hunk 3 above)
Comment (julianrocha) [reply bignum.h:58]: Could be added if needed, shouldn't be too bad to implement. I focussed on the operators in the square/multiply algo

File: bignum.h (hunk 4 above)
Comment (julianrocha) [reply bignum.h:55]: We didn't need it yet so I haven't included

**Pull request merged**